# SIMPLE LINEAR REGRESSION(SLR)

### This is an algorithm which is used for prediction.It has two types of variables:Independent and dependent variables.
### We are trying to predict dependent variable based on one independent variable.If the number of independent variableare more than one then it is known as MLR or Multiple Linear Regression.So the basic difference between the two techniques is the number of independent variables.
## If it is one then SLR ,If greater than one then MLR.
## The name of the dataset is Hotel.csv and is imported as a dataframe in python.
### The dependent variable is that which you want to predict.In our case it is revenue(for hotel file).The independent variable is the one which you are using to predict the dependent variable.In our case it is occupancy. Create a linear regression model to calculate the dependent variable(dv) and occupancy(iv).
## Step1. Import the pandas library.
## Step2. Import the dataset.
## Step3. Create two dataframes: one to store the dependent and the other to store independent variables. Suppose df is the original dataframe.

In [67]:
import pandas as pd

In [68]:
df=pd.read_csv('F:/Hotel.csv')

In [69]:
df.head()

Revenueindollar1000  PercentOccupancy
0               514.44              65.7
1               463.12              61.1
2               598.18              78.2
3               454.92              65.4
4               453.80              63.5

In [70]:
df.tail()

Revenueindollar1000  PercentOccupancy
7                498.70              72.0
8                514.46              72.9
9                623.29              81.7
10               454.77              62.1
11               385.57              53.4

In [71]:
df.describe()

Revenueindollar1000  PercentOccupancy
count            12.000000         12.000000
mean            507.478333         68.983333
std              74.590546          8.681310
min             385.570000         53.400000
25%             454.882500         63.150000
50%             500.465000         68.150000
75%             535.390000         74.225000
max             626.260000         81.700000

In [72]:
import statsmodels.api as sm

In [73]:
X=df['PercentOccupancy'] # The dataframe X is storing the independent variable which is occupancy.
Y=df['Revenueindollar1000'] # The dataframe Y which is storing the dependent variable which in our case in revenue.

##### To the X dataframe ,we need to add a constant for making a linear model.

In [74]:
X=sm.add_constant(X) # add_constant is the function which is available in statsmodels library.

# To Create the SLR regression model,the python code is given below.

In [75]:
model=sm.OLS(Y,X).fit() #OLS is also a function in statsmodels
model.summary() #To see the summary of the model.

C:\ProgramData\Anaconda3\lib\site-packages\scipy\stats\_stats_py.py:1769: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary.Summary'>
"""
                             OLS Regression Results                            
===============================================================================
Dep. Variable:     Revenueindollar1000   R-squared:                       0.918
Model:                             OLS   Adj. R-squared:                  0.910
Method:                  Least Squares   F-statistic:                     111.8
Date:                 Fri, 14 Apr 2023   Prob (F-statistic):           9.52e-07
Time:                         17:51:57   Log-Likelihood:                -53.252
No. Observations:                   12   AIC:                             110.5
Df Residuals:                       10   BIC:                             111.5
Df Model:                            1                                         
Covariance Type:             nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const              -60.3747     54.100     -1.116      0.291    -180.916      60.167
PercentOccupancy     8.2317      0.779     10.572      0.000       6.497       9.967
==============================================================================
Omnibus:                        1.318   Durbin-Watson:                   1.482
Prob(Omnibus):                  0.517   Jarque-Bera (JB):                0.765
Skew:                          -0.140   Prob(JB):                        0.682
Kurtosis:                       1.795   Cond. No.                         581.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# Y(Predicted Variable)=Bo+B1X1 where Y is dv and X is Iv.
### using the model summary,the equation for predicting revenue is:
## PredictedRevenue=(B0)+B1*occupancy
### The value of B0 and B1 can be obtained from the model summary.
### B0=const coef value;B1=PercentOccupancy coeff value
#### PredictedRevenue=(-60.3)+8.2*Occupancy;occupancy=20%,40%,51.2%,etc.
#### From the above equation we can predict what will be the revenue for any particular occupancy.

# How to find the accuracy of the model:-
## To calculate the error in the model: Smaller the error,more accurate the model.
### Create a new variable in python pre_revenue which will store the predicted revenue for every case in the dataset.There should be 12 predicted revenues.
### create a new variable error=Actual reveue-Pre_revenue(12 values)
### Create a new variable rel_error=Error/original revenue(12 values)
### calculate the mean of these 12 values.
### Since there are some negative values ,use the abs() fuction to convert it into positive.

In [76]:
df['pre_revenue']=(-60.3)+(8.2317*df['PercentOccupancy'])

In [77]:
df.head()

Revenueindollar1000  PercentOccupancy  pre_revenue
0               514.44              65.7    480.52269
1               463.12              61.1    442.65687
2               598.18              78.2    583.41894
3               454.92              65.4    478.05318
4               453.80              63.5    462.41295

In [78]:
df['error']=abs(df['Revenueindollar1000']-df['pre_revenue'])

In [79]:
df.head()

Revenueindollar1000  PercentOccupancy  pre_revenue     error
0               514.44              65.7    480.52269  33.91731
1               463.12              61.1    442.65687  20.46313
2               598.18              78.2    583.41894  14.76106
3               454.92              65.4    478.05318  23.13318
4               453.80              63.5    462.41295   8.61295

In [80]:
df['rel_error']=abs(df['error']/df['Revenueindollar1000'])

In [81]:
df.head()

Revenueindollar1000  PercentOccupancy  pre_revenue     error  rel_error
0               514.44              65.7    480.52269  33.91731   0.065931
1               463.12              61.1    442.65687  20.46313   0.044185
2               598.18              78.2    583.41894  14.76106   0.024677
3               454.92              65.4    478.05318  23.13318   0.050851
4               453.80              63.5    462.41295   8.61295   0.018980

In [82]:
df.describe()

Revenueindollar1000  PercentOccupancy  pre_revenue      error  \
count            12.000000         12.000000    12.000000  12.000000   
mean            507.478333         68.983333   507.550105  18.159475   
std              74.590546          8.681310    71.461943   9.856362   
min             385.570000         53.400000   379.272780   3.881430   
25%             454.882500         63.150000   459.531855  10.448320   
50%             500.465000         68.150000   500.690355  18.386990   
75%             535.390000         74.225000   550.697933  23.682618   
max             626.260000         81.700000   612.229890  33.917310   

       rel_error  
count  12.000000  
mean    0.035840  
std     0.019762  
min     0.008535  
25%     0.018671  
50%     0.033033  
75%     0.049641  
max     0.067540

In [83]:
df.rel_error.mean()*100


3.5839931917033683

## The error is roughly 3.5 percent.Therefore the model is highly accurate in predicting revenue from occupancy.

# Use the file hsb2.csv and create a linear regression model to predict read from the variable math and find the accuracy of the model.


In [84]:
df1=pd.read_csv('hsb2.csv')

In [85]:
df1.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst
0   70       0     4    1       1     1    57     52    41       47     57
1  121       1     4    2       1     3    68     59    53       63     61
2   86       0     4    3       1     1    44     33    54       58     31
3  141       0     4    3       1     3    63     44    47       53     56
4  172       0     4    2       1     2    47     52    57       53     61

In [86]:
df1.head(10)

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst
0   70       0     4    1       1     1    57     52    41       47     57
1  121       1     4    2       1     3    68     59    53       63     61
2   86       0     4    3       1     1    44     33    54       58     31
3  141       0     4    3       1     3    63     44    47       53     56
4  172       0     4    2       1     2    47     52    57       53     61
5  113       0     4    2       1     2    44     52    51       63     61
6   50       0     3    2       1     1    50     59    42       53     61
7   11       0     1    2       1     2    34     46    45       39     36
8   84       0     4    2       1     1    63     57    54       58     51
9   48       0     3    2       1     2    57     55    52       50     51

In [87]:
df1.tail()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst
195   31       1     2    2       2     1    55     59    52       42     56
196  145       1     4    2       1     3    42     46    38       36     46
197  187       1     4    2       2     1    57     41    57       55     52
198  118       1     4    2       1     1    55     62    58       58     61
199  137       1     4    3       1     2    63     65    65       53     61

In [88]:
X1=df1['math'] # The dataframe X1 is storing the independent variable which is math.
Y1=df1['read'] # The dataframe Y1 is storing the dependent variable which is read.

In [89]:
X1=sm.add_constant(X1)

model=sm.OLS(Y1,X1).fit() #OLS is also a function in statsmodels
model.summary() #To see the summary of the model.

In [90]:
df1['pre_value_read']=(14.0725)+0.7248*df1['math'] #B0=14.0725,B1=0.7248

In [91]:
df1.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_read  
0         43.7893  
1         52.4869  
2         53.2117  
3         48.1381  
4         55.3861

In [92]:
df1['error']=abs(df1['read']-df1['pre_value_read'])

In [93]:
df1.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_read    error  
0         43.7893  13.2107  
1         52.4869  15.5131  
2         53.2117   9.2117  
3         48.1381  14.8619  
4         55.3861   8.3861

In [94]:
df1['rel_error']=abs(df1['error']/df1['read'])

In [95]:
df1.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_read    error  rel_error  
0         43.7893  13.2107   0.231767  
1         52.4869  15.5131   0.228134  
2         53.2117   9.2117   0.209357  
3         48.1381  14.8619   0.235903  
4         55.3861   8.3861   0.178428

In [96]:
df1.describe()

id     Gender        race         ses      schtyp        prog  \
count  200.000000  200.00000  200.000000  200.000000  200.000000  200.000000   
mean   100.500000    0.54500    3.430000    2.055000    1.160000    2.025000   
std     57.879185    0.49922    1.039472    0.724291    0.367526    0.690477   
min      1.000000    0.00000    1.000000    1.000000    1.000000    1.000000   
25%     50.750000    0.00000    3.000000    2.000000    1.000000    2.000000   
50%    100.500000    1.00000    4.000000    2.000000    1.000000    2.000000   
75%    150.250000    1.00000    4.000000    3.000000    1.000000    2.250000   
max    200.000000    1.00000    4.000000    3.000000    2.000000    3.000000   

             read       write        math     science       socst  \
count  200.000000  200.000000  200.000000  200.000000  200.000000   
mean    52.230000   52.775000   52.645000   51.850000   52.405000   
std     10.252937    9.478586    9.368448    9.900891   10.735793   
min     28.000000   31.000000   33.000000   26.000000   26.000000   
25%     44.000000   45.750000   45.000000   44.000000   46.000000   
50%     50.000000   54.000000   52.000000   53.000000   52.000000   
75%     60.000000   60.000000   59.000000   58.000000   61.000000   
max     76.000000   67.000000   75.000000   74.000000   71.000000   

       pre_value_read       error   rel_error  
count      200.000000  200.000000  200.000000  
mean        52.229596    6.073543    0.121104  
std          6.790251    4.684125    0.103239  
min         37.990900    0.064500    0.001500  
25%         46.688500    2.481100    0.046390  
50%         51.762100    4.812500    0.095319  
75%         56.835700    9.043850    0.173864  
max         68.432500   23.559500    0.615675

In [97]:
df1.rel_error.mean()*100

12.110381214889617

## The error is roughly 12.11 percent.Therefore the model is highly accurate in predicting the variable read from the variable math.

# Make a linear regression to predict write from the variable science and find the accuracy of the model.

In [98]:
df2=pd.read_csv('hsb2.csv')

In [99]:
X2=df2['science'] # The dataframe X1 is storing the independent variable which is science.
Y2=df2['write'] # The dataframe Y1 is storing the dependent variable which is write.

In [100]:
X2=sm.add_constant(X2)

In [101]:
model2=sm.OLS(Y2,X2).fit()
model2.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  write   R-squared:                       0.325
Model:                            OLS   Adj. R-squared:                  0.322
Method:                 Least Squares   F-statistic:                     95.51
Date:                Fri, 14 Apr 2023   Prob (F-statistic):           1.17e-18
Time:                        17:51:58   Log-Likelihood:                -693.73
No. Observations:                 200   AIC:                             1391.
Df Residuals:                     198   BIC:                             1398.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         24.4592      2.949      8.293      0.000      18.643      30.276
science        0.5461      0.056      9.773      0.000       0.436       0.656
==============================================================================
Omnibus:                        7.131   Durbin-Watson:                   1.707
Prob(Omnibus):                  0.028   Jarque-Bera (JB):                7.223
Skew:                          -0.466   Prob(JB):                       0.0270
Kurtosis:                       2.997   Cond. No.                         282.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [102]:
df2['pre_value_write']=(24.4592)+0.5461*df2['science'] #B0=24.4592,B1=0.5461

In [103]:
df2.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_write  
0          50.1259  
1          58.8635  
2          56.1330  
3          53.4025  
4          53.4025

In [104]:
df2['error']=abs(df2['write']-df2['pre_value_write'])

In [105]:
df2.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_write    error  
0          50.1259   1.8741  
1          58.8635   0.1365  
2          56.1330  23.1330  
3          53.4025   9.4025  
4          53.4025   1.4025

In [106]:
df2['rel_error']=abs(df2['error']/df2['write'])

In [107]:
df2.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_write    error  rel_error  
0          50.1259   1.8741   0.036040  
1          58.8635   0.1365   0.002314  
2          56.1330  23.1330   0.701000  
3          53.4025   9.4025   0.213693  
4          53.4025   1.4025   0.026971

In [108]:
df2.describe()

id     Gender        race         ses      schtyp        prog  \
count  200.000000  200.00000  200.000000  200.000000  200.000000  200.000000   
mean   100.500000    0.54500    3.430000    2.055000    1.160000    2.025000   
std     57.879185    0.49922    1.039472    0.724291    0.367526    0.690477   
min      1.000000    0.00000    1.000000    1.000000    1.000000    1.000000   
25%     50.750000    0.00000    3.000000    2.000000    1.000000    2.000000   
50%    100.500000    1.00000    4.000000    2.000000    1.000000    2.000000   
75%    150.250000    1.00000    4.000000    3.000000    1.000000    2.250000   
max    200.000000    1.00000    4.000000    3.000000    2.000000    3.000000   

             read       write        math     science       socst  \
count  200.000000  200.000000  200.000000  200.000000  200.000000   
mean    52.230000   52.775000   52.645000   51.850000   52.405000   
std     10.252937    9.478586    9.368448    9.900891   10.735793   
min     28.000000   31.000000   33.000000   26.000000   26.000000   
25%     44.000000   45.750000   45.000000   44.000000   46.000000   
50%     50.000000   54.000000   52.000000   53.000000   52.000000   
75%     60.000000   60.000000   59.000000   58.000000   61.000000   
max     76.000000   67.000000   75.000000   74.000000   71.000000   

       pre_value_write       error   rel_error  
count       200.000000  200.000000  200.000000  
mean         52.774485    6.127621    0.127428  
std           5.406876    4.782441    0.122271  
min          38.657800    0.136500    0.002260  
25%          48.487600    2.133000    0.039934  
50%          53.402500    5.539000    0.094873  
75%          56.133000    8.867000    0.165094  
max          64.870600   23.133000    0.701000

In [109]:
df2.rel_error.mean()*100

12.742789050810394

## The error is roughly 12.7 percent.Therefore the model is highly accurate in predicting the variable write from the variable science.

# TO predict read from the variable write.


In [110]:
df3=pd.read_csv('hsb2.csv')

In [111]:
X3=df3['write']
Y3=df3['read']

In [112]:
X3=sm.add_constant(X3)

In [113]:
model3=sm.OLS(Y3,X3).fit()
model3.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   read   R-squared:                       0.356
Model:                            OLS   Adj. R-squared:                  0.353
Method:                 Least Squares   F-statistic:                     109.5
Date:                Fri, 14 Apr 2023   Prob (F-statistic):           1.11e-20
Time:                        17:51:58   Log-Likelihood:                -704.77
No. Observations:                 200   AIC:                             1414.
Df Residuals:                     198   BIC:                             1420.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         18.1622      3.307      5.492      0.000      11.640      24.684
write          0.6455      0.062     10.465      0.000       0.524       0.767
==============================================================================
Omnibus:                        3.237   Durbin-Watson:                   1.909
Prob(Omnibus):                  0.198   Jarque-Bera (JB):                2.414
Skew:                           0.114   Prob(JB):                        0.299
Kurtosis:                       2.512   Cond. No.                         304.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [114]:
df3['pre_value_read']=(18.1622)+0.6455*df3['write']

In [115]:
df3.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_read  
0         51.7282  
1         56.2467  
2         39.4637  
3         46.5642  
4         51.7282

In [116]:
df3['error']=abs(df3['read']-df3['pre_value_read'])

In [117]:
df3.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_read    error  
0         51.7282   5.2718  
1         56.2467  11.7533  
2         39.4637   4.5363  
3         46.5642  16.4358  
4         51.7282   4.7282

In [118]:
df3['rel_error']=abs(df3['error']/df3['read'])

In [119]:
df3.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_read    error  rel_error  
0         51.7282   5.2718   0.092488  
1         56.2467  11.7533   0.172843  
2         39.4637   4.5363   0.103098  
3         46.5642  16.4358   0.260886  
4         51.7282   4.7282   0.100600

In [120]:
df3.rel_error.mean()*100

13.530483976983993

# To predict math from the science variable. 

In [121]:
df4=pd.read_csv('hsb2.csv')

In [122]:
X4=df4['science']
Y4=df4['math']

In [123]:
X4=sm.add_constant(X4)

In [124]:
model4=sm.OLS(Y4,X4).fit()
model4.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   math   R-squared:                       0.398
Model:                            OLS   Adj. R-squared:                  0.395
Method:                 Least Squares   F-statistic:                     130.8
Date:                Fri, 14 Apr 2023   Prob (F-statistic):           1.39e-23
Time:                        17:51:59   Log-Likelihood:                -680.04
No. Observations:                 200   AIC:                             1364.
Df Residuals:                     198   BIC:                             1371.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         21.7002      2.754      7.879      0.000      16.269      27.132
science        0.5968      0.052     11.437      0.000       0.494       0.700
==============================================================================
Omnibus:                        1.407   Durbin-Watson:                   2.126
Prob(Omnibus):                  0.495   Jarque-Bera (JB):                1.053
Skew:                          -0.135   Prob(JB):                        0.591
Kurtosis:                       3.230   Cond. No.                         282.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [125]:
df4['pre_value_math']=(21.7002)+0.5968*df4['science']

In [126]:
df4.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_math  
0         49.7498  
1         59.2986  
2         56.3146  
3         53.3306  
4         53.3306

In [127]:
df4['error']=abs(df4['math']-df4['pre_value_math'])

In [128]:
df4.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_math   error  
0         49.7498  8.7498  
1         59.2986  6.2986  
2         56.3146  2.3146  
3         53.3306  6.3306  
4         53.3306  3.6694

In [129]:
df4['rel_error']=abs(df4['error']/df4['math'])

In [130]:
df4.head()

id  Gender  race  ses  schtyp  prog  read  write  math  science  socst  \
0   70       0     4    1       1     1    57     52    41       47     57   
1  121       1     4    2       1     3    68     59    53       63     61   
2   86       0     4    3       1     1    44     33    54       58     31   
3  141       0     4    3       1     3    63     44    47       53     56   
4  172       0     4    2       1     2    47     52    57       53     61   

   pre_value_math   error  rel_error  
0         49.7498  8.7498   0.213410  
1         59.2986  6.2986   0.118842  
2         56.3146  2.3146   0.042863  
3         53.3306  6.3306   0.134694  
4         53.3306  3.6694   0.064375

In [131]:
df4.describe()

id     Gender        race         ses      schtyp        prog  \
count  200.000000  200.00000  200.000000  200.000000  200.000000  200.000000   
mean   100.500000    0.54500    3.430000    2.055000    1.160000    2.025000   
std     57.879185    0.49922    1.039472    0.724291    0.367526    0.690477   
min      1.000000    0.00000    1.000000    1.000000    1.000000    1.000000   
25%     50.750000    0.00000    3.000000    2.000000    1.000000    2.000000   
50%    100.500000    1.00000    4.000000    2.000000    1.000000    2.000000   
75%    150.250000    1.00000    4.000000    3.000000    1.000000    2.250000   
max    200.000000    1.00000    4.000000    3.000000    2.000000    3.000000   

             read       write        math     science       socst  \
count  200.000000  200.000000  200.000000  200.000000  200.000000   
mean    52.230000   52.775000   52.645000   51.850000   52.405000   
std     10.252937    9.478586    9.368448    9.900891   10.735793   
min     28.000000   31.000000   33.000000   26.000000   26.000000   
25%     44.000000   45.750000   45.000000   44.000000   46.000000   
50%     50.000000   54.000000   52.000000   53.000000   52.000000   
75%     60.000000   60.000000   59.000000   58.000000   61.000000   
max     76.000000   67.000000   75.000000   74.000000   71.000000   

       pre_value_math       error   rel_error  
count      200.000000  200.000000  200.000000  
mean        52.644280    5.736152    0.113045  
std          5.908852    4.447793    0.098013  
min         37.217000    0.008600    0.000205  
25%         47.959400    2.173000    0.042863  
50%         53.330600    4.967400    0.094010  
75%         56.314600    8.338900    0.153028  
max         65.863400   26.089000    0.745400

In [132]:
df4.rel_error.mean()*100 

11.304473939653592

# Correlation Function

In [133]:
from scipy.stats import pearsonr

In [134]:
corr,_=pearsonr(df['PercentOccupancy'],df['Revenueindollar1000'])

In [135]:
corr #Higher the correlation,lower the error...

0.9580612491279302